jax/_src/scipy/cluster/vq.py

```diff
@@ -0,0 +1,41 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import operator
+
+import scipy.cluster.vq
+import textwrap
+
+from jax import vmap
+from jax._src.numpy.util import _wraps, _promote_args_inexact
+from jax._src.numpy.lax_numpy import argmin
+from jax._src.numpy.linalg import norm
+
+
+_no_chkfinite_doc = textwrap.dedent("""
+Does not currently support the Scipy argument ``check_finite=True``,
+because `jax.numpy.asarray_chkfinite` does not exist at the moment.
+""")
+
+
+@_wraps(scipy.cluster.vq.vq, lax_description=_no_chkfinite_doc, skip_params=('check_finite',))
+def vq(obs, code_book, check_finite=True):
+    # note: _promote_args_inexact perform _check_arraylike and broadcasting
+    obs, code_book = _promote_args_inexact("vq", obs, code_book)

```

Comment:
This seems off to me. In general `obs` and `code_book` may have different leading dimensions.

Also, I believe we should be erroring if obs and code_book do not have the same number of dimensions, or if they are 2D and do not have the same second dimension.

Reply:
Great catch, I will dig into `scipy` behavior and fix it. I will strengthen the tests cases as well.

Reply:
Okay, I misunderstand the effect of `_promote_shapes`.
Its "docstr" is confusing:
"Apply NumPy-style broadcasting, making args shape-compatible for lax.py."
But actually it do nothing if args have same rank.

Reply:
Agreed, that's confusing, but I don't think we should be calling `_promote_shapes` at all. The inputs to this function are explicitly not broadcast-compatible.

Reply:
Yes I explicitly call `_check_arraylike` and use `_promote_dtypes_inexact` now.